### PR TITLE
NCCL API Executor Integration

### PR DESCRIPTION
**Pull request by caiomcbr**

### CODE REVIEW

File: apps/nccl/src/nccl.cu [hunk 1]
```diff
@@ -151,6 +155,86 @@ static std::shared_ptr<mscclpp::DeviceHandle<mscclpp::SmChannel>> setupSmChannel
   return ptr;
 }
 
+static ncclResult_t ncclAllReduceOld(const void* sendbuff, void* recvbuff, size_t count, ncclDataType_t datatype,
```
Comment (chhwang) [inline apps/nccl/src/nccl.cu:158]: Please name this function as `ncclAllReduceFallback`

File: apps/nccl/src/nccl.cu [hunk 2]
```diff
@@ -321,82 +419,45 @@ NCCL_API ncclResult_t ncclBroadcast(const void*, void*, size_t, ncclDataType_t,
...
-    auto sendIt = comm->channelScratchInfos.find(sendKey);
-    if (sendIt == comm->channelScratchInfos.end()) {
-      std::vector<mscclpp::SmChannel> channels =
-          setupSmChannels(comm, comm->remoteScratchRegMemories, const_cast<void*>((void*)sendBasePtr));
-      ChannelInfo channelInfo{channels, setupSmChannelDeviceHandles(channels)};
-      sendIt = comm->channelScratchInfos.emplace(sendKey, channelInfo).first;
-    }
-
-    smChannels = sendIt->second.smChannelDeviceHandles.get();
+  if (bytes < 16 * (1 << 10)) {
```
Comment (chhwang) [inline apps/nccl/src/nccl.cu:431]: Can we make this configurable by users?

File: apps/nccl/src/nccl.cu [hunk 3]
```diff
@@ -321,82 +419,45 @@ NCCL_API ncclResult_t ncclBroadcast(const void*, void*, size_t, ncclDataType_t,
...
-    auto recvIt = comm->channelOutInfos.find(recvKey);
-    if (recvIt == comm->channelOutInfos.end()) {
-      remoteMemories =
-          setupRemoteMemories(comm->comm, rank, (void*)recvBasePtr, recvBytes, mscclpp::Transport::CudaIpc);
-      std::vector<mscclpp::SmChannel> outChannels =
-          setupSmChannels(comm, remoteMemories, const_cast<void*>((void*)recvBasePtr));
-      ChannelInfo channelInfo{outChannels, setupSmChannelDeviceHandles(outChannels)};
-      recvIt = comm->channelOutInfos.emplace(recvKey, channelInfo).first;
+    std::shared_ptr<mscclpp::ExecutionPlan> plan;
+    if (bytes <= (1 << 20))
```
Comment (chhwang) [inline apps/nccl/src/nccl.cu:435]: Can we make this configurable by users?

**Review by chhwang — CHANGES_REQUESTED** [review]

File: apps/nccl/src/nccl.cu (hunk 3 above)
Comment (caiomcbr) [reply apps/nccl/src/nccl.cu:435]: Do you mean give the user the option to choose between the allReduce using the executor and the allReduceFallBack?

Comment (caiomcbr) [reply apps/nccl/src/nccl.cu:435]: Do you mean give the user the option to choose the byte range to use each json file, right?

Comment (chhwang) [reply apps/nccl/src/nccl.cu:435]: > Do you mean give the user the option to choose the byte range to use each json file, right?

Yes, right

File: apps/nccl/src/nccl.cu [hunk 4]
```diff
@@ -151,6 +195,86 @@ static std::shared_ptr<mscclpp::DeviceHandle<mscclpp::SmChannel>> setupSmChannel
...
+  uint32_t scratchBuffIdx = (++(comm->buffFlag)) % comm->numScratchBuff;
+  size_t offsetScratch = (SCRATCH_SIZE / comm->numScratchBuff) * scratchBuffIdx;
+  int rank = comm->comm->bootstrap()->getRank();
+  channelKey sendKey{(void*)sendBasePtr, sendBytes};
+  channelKey recvKey{(void*)recvBasePtr, recvBytes};
+  mscclpp::DeviceHandle<mscclpp::SmChannel>* smChannels = nullptr;
+  mscclpp::DeviceHandle<mscclpp::SmChannel>* smOutChannels = nullptr;
+
+  // Creating the channels
+  if (count * ncclTypeSize(datatype) <= (1 << 20)) {
```
Comment (chhwang) [inline apps/nccl/src/nccl.cu:220]: Should we use `ALLREDUCE_LARGE_MSG_BOUNDARY` here as well?

**Review by chhwang — APPROVED** [review]

**Pull request merged**